p2p/src/net/libp2p/mod.rs

```diff
@@ -208,23 +270,62 @@ impl NetworkService for Libp2pService {
             .boxed();
 
         let swarm = {
-            let mut behaviour = common::ComposedBehaviour {
-                streaming: Streaming::<IdentityCodec>::default(),
+            let gossipsub_config = GossipsubConfigBuilder::default()
+                .heartbeat_interval(std::time::Duration::from_secs(10))
+                .validation_mode(ValidationMode::Strict)
+                .max_transmit_size(2 * 1024 * 1024)
+                .validate_messages()
+                .build()
+                .expect("configuration to be valid");
+
+            // TODO: impl display for semver/magic bytes?
+            let version = config.version();
+            let magic = config.magic_bytes();
+            let protocol = format!(
+                "/mintlayer/{}.{}.{}-{:x}",
+                version.major,
+                version.minor,
+                version.patch,
+                ((magic[0] as u32) << 24)
+                    | ((magic[1] as u32) << 16)
+                    | ((magic[2] as u32) << 8)
+                    | (magic[3] as u32)
+            );
+
+            let mut behaviour = types::ComposedBehaviour {
                 mdns: Mdns::new(Default::default()).await?,
-                floodsub: Floodsub::new(peer_id),
+                ping: ping::Behaviour::new(
+                    ping::Config::new()
+                        .with_timeout(std::time::Duration::from_secs(60))
+                        .with_interval(std::time::Duration::from_secs(60))
+                        .with_max_failures(NonZeroU32::new(3).expect("max failures > 0")),
+                ),
+                identify: Identify::new(IdentifyConfig::new(protocol, id_keys.public())),
+                sync: RequestResponse::new(
+                    SyncingCodec(),
+                    iter::once((SyncingProtocol(), ProtocolSupport::Full)),
+                    RequestResponseConfig::default(),
+                ),
+                gossipsub: Gossipsub::new(
+                    MessageAuthenticity::Signed(id_keys.clone()),
+                    gossipsub_config,
+                )
+                .expect("configuration to be valid"),
             };
 
             for topic in topics.iter() {
-                log::debug!("subscribing to floodsub topic {:?}", topic);
-                behaviour.floodsub.subscribe(topic.into());
+                log::debug!("subscribing to gossipsub topic {:?}", topic);
+                behaviour.gossipsub.subscribe(&topic.into()).expect("subscription to work");
             }
 
+            // subscribes to our topic
             SwarmBuilder::new(transport, behaviour, peer_id).build()
         };
 
         let (cmd_tx, cmd_rx) = mpsc::channel(16);
-        let (flood_tx, flood_rx) = mpsc::channel(64);
+        let (gossip_tx, gossip_rx) = mpsc::channel(64);
         let (conn_tx, conn_rx) = mpsc::channel(64);
+        let (sync_tx, sync_rx) = mpsc::channel(64);
```

Comment:
It would make sense but the fate of these channels are .

I also pick just some nice power of two that's large enough but I think the most appropriate way would be to take measurements (to find distributions for service and arrival times) and then pick appropriate queuing model as there's quite a bit of research on queuing theory. Maybe someday we'll do that